docs/cudf/source/developer_guide/ARCHITECTURE.md

```diff
@@ -0,0 +1,208 @@
+2. The Column layer: The core internal data structures used to bridge the gap to our lower-level implementations.
+3. The Cython layer: The wrappers around the fast C++ `libcudf` library.
+
+In this document we will review each of these layers, their roles, and the requisite tradeoffs.
+
+
+## The Frame layer
+
+Broadly speaking, the `Frame` layer is composed of two types of objects: indexed tables and indexes.
+The mapping between these types and cuDF data types is not obvious, however.
+To ease our way into understanding why, let's first take a birds-eye view of the Frame layer.
+
+All classes in this layer inherit from one or both of the two base classes in this layer: `Frame` and `BaseIndex`.
+The eponymous `Frame` class is, at its core, a simple tabular data structure composed of columnar data.
+Some types of `Frame` contain indexes; in particular, any `DataFrame` or `Series` has an index.
+However, as a general container of columnar data, `Frame` is also the parent class for most types of index.
+
+`BaseIndex`, meanwhile, is essentially an abstract base class encoding the `pandas.Index` API.
+Various subclasses of `BaseIndex` implement this API in specific ways depending on their underlying data.
+For example, `RangeIndex` avoids actually materializing a column, while a `MultiIndex` contains _multiple_ columns.
+Most other index classes consist of a single column of a given type, e.g. strings or datetimes.
+As a result, using a single abstract parent provides the flexibility we need to support these different types.
+
+With those preliminaries out of the way, let's dive in a little bit deeper.
+
+### Frames
+
+`Frame` exposes numerous methods common to all pandas data structures.
+Any methods that have the same API across `Series`, `DataFrame`, and `Index` should be defined here.
+Additionally any (internal) methods that could be used to share code between those classes may also be defined here.
+
+The primary internal subclass of `Frame` is `IndexedFrame`, a `Frame` with an index.
+An `IndexedFrame` represents the first type of object mentioned above: indexed tables.
+In particular, `IndexedFrame` is a parent class for `DataFrame` and `Series`.
+Any pandas methods that are defined for those two classes should be defined here.
+
+The second internal subclass of `Frame` is `SingleColumnFrame`.
+As you may surmise, it is a `Frame` with a single column of data.
+This class is a parent for most types of indexes as well as `Series` (note the diamond inheritance pattern here).
+While `IndexedFrame` provides a large amount of functionality, this class is much simpler.
+It adds some simple APIs provided by all 1D pandas objects, and it flattens outputs where needed.
+
+### Indexes
+
+While we've highlighted some exceptional cases of Indexes before, let's start with the base cases here first.
+`BaseIndex` is generally intended to be a true abstract class, i.e. it should contain no implementations.
```

Comment:
```suggestion
`BaseIndex` is generally intended to be a true abstract class, i.e. it should contain no implementations.
```

Is the definition of a true abstract class that it should not contain any implementations? Or that its (abstract) methods _must_ be overridden?

Reply:
"true abstract class" is not a term I'm familiar with. I'd prefer a simpler sentence overall:
```suggestion
`BaseIndex` is intended to be an abstract class with no implementations.
```

Reply:
I'd be more explicit about the term "abstract" here; maybe something along the lines of:


> `BaseIndex` is intended to behave like an abstract class, i.e., many of its methods simply raise `NotImplementedError` and are intended to be overridden in subclasses.

Reply:
I'm fine with being explicit. For clarification, my wording here should have said "pure" abstract class, which is more standard terminology.